core/src/main/clojure/xtdb/operator/scan.clj

```diff
@@ -71,47 +71,44 @@
+              :param (some-> (-> (.readerForName params (name arg))
+                                 (.getObject 0))
                              (time/sql-temporal->micros (.getZone expr/*clock*)))
-                :param (some-> (-> (.readerForName params (name arg))
-                                   (.getObject 0))
-                               (time/sql-temporal->micros (.getZone expr/*clock*)))
-                :now (-> (.instant expr/*clock*)
-                         (time/instant->micros))))]
+              :now (-> (.instant expr/*clock*)
+                       (time/instant->micros))))
+          (apply-constraint [constraint]
+            (if-let [[tag & args] constraint]
+              (case tag
+                :at (let [[at] args
+                          at-μs (->time-μs at)]
+                      (TemporalDimension/at at-μs))
 
-      (when-let [system-time (some-> at-tx (.getSystemTime) time/instant->micros)]
-        (.lte (.getSystemFrom bounds) system-time)
+                ;; overlaps [time-from time-to]
+                :in (let [[from to] args]
+                      (TemporalDimension/in (->time-μs (or from [:now]))
+                                            (some-> to ->time-μs)))
 
-        (when-not for-system-time
-          (.gt (.getSystemTo bounds) system-time)))
+                :between (let [[from to] args]
+                           (TemporalDimension/between (->time-μs (or from [:now]))
+                                                      (some-> to ->time-μs)))
 
-      (letfn [(apply-constraint [constraint ^TemporalBounds$TemporalColumn start-col, ^TemporalBounds$TemporalColumn end-col]
-                (when-let [[tag & args] constraint]
-                  (case tag
-                    :at (let [[at] args
-                              at-μs (->time-μs at)]
-                          (.lte start-col at-μs)
-                          (.gt end-col at-μs))
+                :all-time (TemporalDimension.))
+              (TemporalDimension.)))]
 
-                    ;; overlaps [time-from time-to]
-                    :in (let [[from to] args]
-                          (.gt end-col (->time-μs (or from [:now])))
-                          (when-let [to-μs (some-> to ->time-μs)]
-                            (.lt start-col to-μs)))
-
-                    :between (let [[from to] args]
-                               (.gt end-col (->time-μs (or from [:now])))
-                               (when-let [to-μs (some-> to ->time-μs)]
-                                 (.lte start-col to-μs)))
+    (let [^TemporalDimension sys-dim (apply-constraint for-system-time)
+          bounds (TemporalBounds. (apply-constraint for-valid-time) sys-dim)]
+      ;; we further constrain bases on tx
+      (when-let [system-time (some-> at-tx (.getSystemTime) time/instant->micros)]
+        (.setMaxTo sys-dim (min (inc system-time) (.getMaxTo sys-dim)))
```

Comment:
This is now the only location where a bit more fiddly logic is required. We first create TemporalBounds based on `for-system-time` (if existant) and then further constrain based on `at-tx`.